lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,217 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# NOTE: We won't always be able to import from snowflake.{connector, snowpark}.* so need
+# the `type: ignore` comment below, but that comment will explode if `warn-unused-ignores`
+# is turned on when the package is available. Unfortunately, mypy doesn't provide a good
+# way to configure this at a per-line level :(
+# mypy: no-warn-unused-ignores
+
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import TYPE_CHECKING, Sequence, cast
+
+import pandas as pd
+
+from streamlit.connections import BaseConnection
+from streamlit.connections.util import load_from_snowsql_config_file, running_in_sis
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.connector import (  # type: ignore # isort: skip
+        SnowflakeConnection as InternalSnowflakeConnection,
+    )
+    from snowflake.connector.cursor import SnowflakeCursor  # type: ignore
+    from snowflake.snowpark.session import Session  # type: ignore
+
+
+_REQUIRED_CONNECTION_PARAMS = {"account"}
+
+
+class SnowflakeConnection(BaseConnection["InternalSnowflakeConnection"]):
+    """A connection to Snowflake using the Snowflake Python Connector. Initialize using
+    ``st.connection("<name>", type="snowflake")``.
+
+    TODO(vdonato): Finish this docstring.
+    """
+
+    def _connect(self, **kwargs) -> "InternalSnowflakeConnection":
+        import snowflake.connector  # type: ignore
+        from snowflake.snowpark.context import get_active_session  # type: ignore
+
+        # If we're running in SiS, just call get_active_session() and retrieve the
+        # lower-level connection from it.
+        if running_in_sis():
+            session = get_active_session()
+
+            if hasattr(session, "connection"):
+                return session.connection
+            return session._conn._conn
+
+        # Otherwise, attempt to create a new connection from whatever credentials we
+        # have available.
+
+        st_secrets = self._secrets.to_dict()
+        if len(st_secrets):
+            # TODO(vdonato): Double-check that we want to merge any secrets we find with
+            # the kwargs passed to `st.connection`.
+            conn_kwargs = {**st_secrets, **kwargs}
+            return snowflake.connector.connect(**conn_kwargs)
+
+        if hasattr(snowflake.connector.connection, "CONFIG_MANAGER"):
+            # TODO(vdonato): Double-check that we want to map the "snowflake" connection
+            # name in a call to `st.connection` to the "default" connection in the
+            # Snowflake config file. This is aligned with the default behavior for the
+            # underlying python connector and seems more sane than having a "snowflake"
+            # connection in a snowflake-specific config file.
```

Comment:
btw @sfc-gh-jcarroll these two `TODO`s could use a bit of product input.

I'm fairly certain that both implementations here are pretty reasonable (I think we'll want `kwargs` to apply to any method of creating a `SnowflakeConnection`, and mapping `snowflake` -> `default` seems the most natural to me), but it's worth getting some input on this

Reply:
The other thing to consider is whether we want to map the `snowflake` connection name to `default` as well on L93

Reply:
Yeah for the above 2 TODOs that looks correct to me as well. We should get confirmation from Mark that it seems consistent to him too.

For L93 I'm not totally sure I follow, maybe we can discuss on Monday.